app/id_generation.py

```diff
@@ -2,41 +2,142 @@
 import requests
 from settings import settings
 import random
+from helpers import db_request_token, is_response_valid, is_response_empty
+from request import build_request
+from router import (
+    user as user_router,
+    student as student_router,
+    school,
+    enrollment_record,
+)
 
 school_db_url = settings.db_url + "/school"
 
 
 class JNVIDGeneration:
-    def __init__(self, region, school, grade):
-        self.region = region
-        self.school = school
-        self.grade = grade
-        self.id = (
-            self.get_class_code()
-            + self.get_jnv_code()
-            + self.generate_three_digit_code()
+    def __init__(self, parameters):
+        self.parameters = parameters
+
+        record_already_exist, ID = self.dedupe_for_users()
+        if record_already_exist:
+            self.id = ID
+
+        else:
+            counter = 1000
+            while counter > 0:
+                id = (
+                    self.get_class_code()
+                    + self.get_jnv_code()
+                    + self.generate_three_digit_code()
+                )
+
+                if not self.check_for_duplicate_ID(id):
+                    self.id = id
+                    break
+                counter -= 1
+
+    def dedupe_for_users(self):
+
+        # check if the keys used for deduping exist in the request data
+        for key in ["grade", "date_of_birth", "school_name", "gender", "category"]:
+            if (
+                key not in self.parameters
+                or self.parameters[key] == ""
+                or self.parameters[key] is None
+            ):
+                raise HTTPException(
+                    status_code=400,
+                    detail="{} is not part of the request data".format(key),
+                )
+
+        does_user_already_exist = user_router.get_users(
+            build_request(
+                query_params={
+                    "date_of_birth": self.parameters["date_of_birth"],
+                    "gender": self.parameters["gender"],
+                }
+            )
         )
+        if len(does_user_already_exist) == 0:
+            # if user does not exist, go ahead with ID generation
+            return [False, ""]
+
+        else:
+            # if user does exist, check corresponding student attributes
+            for user in does_user_already_exist:
```

Comment:
maybe call it `get_user_response` instead of `does_user_already_exist`, which sounds like a boolean.